### ISSUE TITLE
[apmgrpc] Call other service methods in goroutine, prompting error: context canceled

### ISSUE THREAD

**Issue by EDDYCJY** [elastic/apm-agent-go#475]
**Describe the bug**
In the A method, the B method is called by the goroutine, and rpc error occurs: code = Canceled desc = context canceled

**Code**
```
...
func (k *Server) Get(ctx context.Context, r *pb.Request) (*pb.Response, error) {
	go func() {
		resp, err := CallBServerMethod(ctx)
		if err != nil {
			log.Println(err)
		}
	}()
        ... 
}
```

**To Reproduce**
1. 2 gRPC Servers, A and B.
2. The RPC method in A Server, call the RPC method of B Server via goroutine (pass ctx for link tracking)
3. The RPC method call of A Server ends, and the RPC method of B Server returned by goroutine returns an error.
4. See error: `rpc error: code = Canceled desc = context canceled`

**Expected behavior**
Call is normal, link tracking is normal.

--

There is a problem here. If ctx is not passed, the link tracking is incomplete. However, if ctx is passed, an error will occur in the scene called by goroutine.

What should I do?

**Comment by axw** [comment]
The problem -- which you probably know already, but it bears repeating -- is that the context passed into A is cancelled once A returns.

What you'll need to do is pass in a new context, copying across the https://godoc.org/go.elastic.co/apm#Transaction and https://godoc.org/go.elastic.co/apm#Span if any. This is a little bit cumbersome at the moment, but here's how you can do that:

```go
newctx := context.Background()
if tx := apm.TransactionFromContext(ctx); tx != nil {
    newctx = apm.ContextWithTransaction(newctx, tx)
}
if span := apm.SpanFromContext(ctx); span != nil {
    newctx = apm.ContextWithTransaction(newctx, span)
}
```

Then pass `newctx` into `CallBServerMethod`.

I'll create an issue to make this easier.

**Comment by EDDYCJY** [comment]
@axw Thanks for the reply, I am trying Demo, but I have a problem:

```
if span := apm.SpanFromContext(ctx); span != nil {
    newctx = apm.ContextWithTransaction(newctx, span)
}
```

Span is *Span, apm.ContextWithTransaction entry is *Transaction

ps.
```
...
func SpanFromContext(ctx context.Context) *Span {
	value, _ := apmcontext.SpanFromContext(ctx).(*Span)
	return value
}

func ContextWithTransaction(parent context.Context, t *Transaction) context.Context {
	return apmcontext.ContextWithTransaction(parent, t)
}
```

**Comment by axw** [comment]
@EDDYCJY sorry that was a typo, I meant to write:

```go
if span := apm.SpanFromContext(ctx); span != nil {
    newctx = apm.ContextWithSpan(newctx, span)
}
```

**Comment by EDDYCJY** [comment]
I tried it, but there are still some problems, the code is as follows:

```
func (k *Server) Get(ctx context.Context, r *pb.Request) (*pb.Response, error) {
        log.Printf("ctx: %+v\n", ctx)

	newctx := context.Background()
	if tx := apm.TransactionFromContext(ctx); tx != nil {
		newctx = apm.ContextWithTransaction(newctx, tx)
	}
	if span := apm.SpanFromContext(ctx); span != nil {
		newctx = apm.ContextWithSpan(newctx, span)
	}

	log.Printf("newctx: %+v\n", newctx)

	go func() {
		resp, err := CallBServerMethod(newctx)
		if err != nil {
			log.Println(err)
		}
	}()
        ... 
}
```
Questions are as follows:

The call links for different RPC requests are still separate and are not shown together

---

ps:
```
2019/03/15 13:38:16 ctx: context.Background.WithCancel.WithValue(&http.contextKey{name:"local-addr"}, &net.TCPAddr{IP:net.IP{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xff, 0xff, 0x7f, 0x0, 0x0, 0x1}, Port:10001, Zone:""}).WithValue(&http.contextKey{name:"http-server"}, &http.Server{Addr:":10001", Handler:(http.HandlerFunc)(0x46bfcf0), TLSConfig:(*tls.Config)(0xc00033aa80), ReadTimeout:0, ReadHeaderTimeout:0, WriteTimeout:0, IdleTimeout:0, MaxHeaderBytes:0, TLSNextProto:map[string]func(*http.Server, *tls.Conn, http.Handler){"h2":(func(*http.Server, *tls.Conn, http.Handler))(0x42d4a00)}, ConnState:(func(net.Conn, http.ConnState))(nil), ErrorLog:(*log.Logger)(nil), disableKeepAlives:0, inShutdown:0, nextProtoOnce:sync.Once{m:sync.Mutex{state:0, sema:0x0}, done:0x1}, nextProtoErr:error(nil), mu:sync.Mutex{state:0, sema:0x0}, listeners:map[*net.Listener]struct {}{(*net.Listener)(0xc000351770):struct {}{}}, activeConn:map[*http.conn]struct {}{(*http.conn)(0xc000274000):struct {}{}, (*http.conn)(0xc000274140):struct {}{}}, doneChan:(chan struct {})(nil), onShutdown:[]func(){(func())(0x42dd9e0)}}).WithCancel.WithCancel.WithValue(metadata.mdIncomingKey{}, metadata.MD{":authority":[]string{"grpc-server"}, "grpcgateway-cache-control":[]string{"no-cache"}, "grpcgateway-user-agent":[]string{"PostmanRuntime/7.1.1"}, "grpcgateway-accept":[]string{"*/*"}, "x-forwarded-host":[]string{"127.0.0.1:10001"}, "x-forwarded-for":[]string{"127.0.0.1"}, "user-agent":[]string{"grpc-go/1.14.0"}, "content-type":[]string{"application/grpc"}}).WithValue(peer.peerKey{}, &peer.Peer{Addr:"127.0.0.1:52344", AuthInfo:credentials.TLSInfo{State:tls.ConnectionState{Version:0x303, HandshakeComplete:true, DidResume:false, CipherSuite:0xc02b, NegotiatedProtocol:"h2", NegotiatedProtocolIsMutual:true, ServerName:"grpc-server", PeerCertificates:[]*x509.Certificate(nil), VerifiedChains:[][]*x509.Certificate(nil), SignedCertificateTimestamps:[][]uint8(nil), OCSPResponse:[]uint8(nil), ekm:(func(string, []uint8, int) ([]uint8, error))(0x42396d0), TLSUnique:[]uint8{0x7e, 0x2, 0xd7, 0x99, 0xb, 0x42, 0xfc, 0x16, 0x1f, 0x92, 0x4a, 0x85}}}}).WithValue(grpc.streamKey{}, <stream: 0xc0004eb200, /kibana.KibanaService1/Get>).WithValue(apmcontext.transactionKey{}, &apm.Transaction{mu:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, TransactionData:(*apm.TransactionData)(0xc00019f800)})
2019/03/15 13:38:16 newctx: context.Background.WithValue(apmcontext.transactionKey{}, &apm.Transaction{mu:sync.RWMutex{w:sync.Mutex{state:0, sema:0x0}, writerSem:0x0, readerSem:0x0, readerCount:0, readerWait:0}, TransactionData:(*apm.TransactionData)(0xc00019f800)})
```

**Comment by EDDYCJY** [comment]
Although the call is normal (because `context. Background ()`), the links are not concatenated. Where is the problem?

**Comment by EDDYCJY** [comment]
now:

![image](https://i.imgur.com/KNBjQ03.jpg)

expect:

![image](https://i.imgur.com/TpNwHLl.jpg)

**Comment by axw** [comment]
I think I see the problem now. The transaction is being ended before method "B" creates the span, which causes the span to be discarded. I'll look into fixing this. I don't think there's any simple workaround.

**Comment by EDDYCJY** [comment]
I can understand the situation, but I think this should be supported because it is very common in Go. Is it possible to consider adding a mechanism that can still complete the span after the transaction ends?

**Comment by axw** [comment]
> I can understand the situation, but I think this should be supported because it is very common in Go. Is it possible to consider adding a mechanism that can still complete the span after the transaction ends?

Yes, I'm investigating a solution now.

**Comment by EDDYCJY** [comment]
```
func (k *Server) Get(ctx context.Context, r *pb.Request) (*pb.Response, error) {
        log.Printf("ctx: %+v\n", ctx)

	newctx := context.Background()
	if tx := apm.TransactionFromContext(ctx); tx != nil {
		newctx = apm.ContextWithTransaction(newctx, tx)
	}
	if span := apm.SpanFromContext(ctx); span != nil {
		newctx = apm.ContextWithSpan(newctx, span)
	}

	log.Printf("newctx: %+v\n", newctx)

	go func() {
		resp, err := CallBServerMethod(newctx)
		if err != nil {
			log.Println(err)
		}
	}()
        ... 
        time.Sleep(time.Second*10)
}
```

In line with the conjecture, if you let transaction A forcefully wait, the span will appear

**Comment by EDDYCJY** [comment]
> > I can understand the situation, but I think this should be supported because it is very common in Go. Is it possible to consider adding a mechanism that can still complete the span after the transaction ends?
> 
> Yes, I'm investigating a solution now.

Ok, then I will use the copy scheme to support the use of production. If there is an update later, I will follow up.

thank

**Comment by axw** [comment]
@EDDYCJY https://github.com/elastic/apm-agent-go/pull/478 should resolve the second issue. This change will allow a span to be created from an ended transaction.

I am still testing the change. I hope to have it merged later this week, with a new release soon after.

**Comment by EDDYCJY** [comment]
@axw Good thanks, I will continue to pay attention.

**Comment by axw** [comment]
@EDDYCJY #478 is merged, so you can try the fix if you update to the master branch.

I've also just created a new PR, https://github.com/elastic/apm-agent-go/pull/481, which introduces a top-level function `apm.DetachedContext`. You can use this to create a context which propagates trace context, but will not be cancelled if the enclosing context is cancelled or its deadline is reached.

So instead of

```go
newctx := context.Background()
if tx := apm.TransactionFromContext(ctx); tx != nil {
    newctx = apm.ContextWithTransaction(newctx, tx)
}
if span := apm.SpanFromContext(ctx); span != nil {
    newctx = apm.ContextWithSpan(newctx, span)
}
```

you would write

```go
newctx := apm.DetachedContext(ctx)
```

I would appreciate your feedback on this.

**Comment by EDDYCJY** [comment]
When is the expected release of the next version? (apm.DetachedContext)

**Comment by axw** [comment]
> When is the expected release of the next version? (apm.DetachedContext)

I hope to create a new release in the next couple of days.

**Comment by axw** [comment]
@EDDYCJY I have just merged DetachedContext into master, so you can try the master branch in the mean time if you like. I'll create the proper release/tags soon.

**Comment by EDDYCJY** [comment]
@axw Ok, understand the situation. Because we are using it in a production environment, it is generally safer to use a released version.

**Comment by EDDYCJY** [comment]
@axw I have already tried it, it is ok. as follows:

![image](http://wx1.sinaimg.cn/large/006fVPCvly1g199mhfpkoj31uo0digo7.jpg)

Thanks, great! Next, I will wait for the official version to be released before updating the production environment.

**Comment by axw** [comment]
Excellent, thanks for testing.

**Comment by axw** [comment]
@EDDYCJY I've just released v1.3.0: https://github.com/elastic/apm-agent-go/releases/tag/v1.3.0. Thanks for raising the issue, and again for testing.
